### PR TITLE
FAI-578: Data Dictionary - PMML editor - integration test

### PR DESCRIPTION
**Pull request by sjamboro**
FAI:578: https://issues.redhat.com/browse/FAI-578

Goal: create test suite which cover user interaction with Data Dictionary editor
Non-goal: check all properties of DF

### CODE REVIEW

**Comment by sjamboro** [comment]
Jenkins retest this

**Comment by sjamboro** [comment]
@manstis @kelvah @tiagobento can you review this PR?

File: packages/pmml-editor/it-tests/integration/data-dictionary.ts [hunk 1]
```diff
@@ -0,0 +1,517 @@
...
+        '  <DataDictionary numberOfFields="2">' +
+          '    <DataField name="aaa" optype="categorical" dataType="string"/>' +
+          '    <DataField name="bbb" optype="categorical" dataType="string"/>' +
+          "  </DataDictionary>"
+      )
+      .ouiaId("pmml-modal-confirm")
+      .click();
+    /*
+     * Current implementation allows use string "test\ntest" without any warning
+     * TODO: discuss this topic with developers
```
Comment (manstis) [inline packages/pmml-editor/it-tests/integration/data-dictionary.ts:74]: @kelvah FYI.

Comment (manstis) [inline packages/pmml-editor/it-tests/integration/data-dictionary.ts:74]: @kelvah FYI.

File: packages/pmml-editor/it-tests/integration/data-dictionary.ts [hunk 2]
```diff
@@ -0,0 +1,517 @@
...
+        /*
+         * Unset property
+         */
+        cy.ouia("edit-props", "link-label").click();
+        cy.ouia("df-props", "editor-body").within(() => {
+          cy.ouiaId("val-0").find("button").click();
+          cy.ouiaId("val-0").find("button").click();
+          cy.ouiaId("val-0").find("button").click();
+          cy.ouiaId("val-0").find("button").click();
+          //PMML Spec says that there can be zero Values, but editor does not allow to delete the last item.
```
Comment (manstis) [inline packages/pmml-editor/it-tests/integration/data-dictionary.ts:316]: @kelvah FYI. @sjamboro and myself were discussing this. We should probably allow deletion of the last item.

File: packages/pmml-editor/it-tests/integration/data-dictionary.ts [hunk 3]
```diff
@@ -0,0 +1,517 @@
...
+        });
+        cy.ouiaId("back-to-DFs").click();
+        cy.ouiaId("dd-toolbar").click();
+
+        cy.ouia("dt-string", "dd-type-item")
+          .within(() => {
+            cy.ouia("Invalid Value", "data-props-label")
+              .should("exist")
+              .should("to.have.text", "Invalid Value:\u00A0invalid");
+            //XSD allows several invalid values. PMML editors allows just one.
```
Comment (manstis) [inline packages/pmml-editor/it-tests/integration/data-dictionary.ts:391]: Yeeeees... and we concluded that the PMML Specification only supports one `Value` with `property=invalid`

**Review by manstis — APPROVED** [review]
LGTM. Thank-you @sjamboro.

I would like to understand why `ouiaId` is used in some places but `data-ouia-component-id` in others and the purpose of `data-ouia-component-type`. Can you please explain?

File: packages/pmml-editor/it-tests/integration/data-dictionary.ts (hunk 1 above)
Comment (kelvah) [reply packages/pmml-editor/it-tests/integration/data-dictionary.ts:74]: I've created a ticket to track it: [FAI-680](https://issues.redhat.com/browse/FAI-680)

File: packages/pmml-editor/it-tests/integration/data-dictionary.ts (hunk 2 above)
Comment (kelvah) [reply packages/pmml-editor/it-tests/integration/data-dictionary.ts:316]: I have to disagree?
I recollect having a similar discussion in the past.
I would treat the two things separately: the spec and what the user sees in the UI. It's true the spec says it's possible to have 0 values defined. But values are used for other properties too (like invalid and missing value).
In our specific scenario that means having no constraints.
When the user wants to add constraints (using the constraint type selection) then he has to provide at least one value. Having the constraints option selected without values makes little sense. Unless we want to automatically clear the constraints type selection when deleting the last value, which could be undesirable when I'm just updating all the values for example.

Comment (manstis) [reply packages/pmml-editor/it-tests/integration/data-dictionary.ts:316]: > I have to disagree?

I don't know; do you have to disagree?

Joking aside. I can see your point: the UI can be "better than the spec" and prevent Users making mistakes. If they've elected to have "Value" constraints, as you say, it makes no sense having zero defined. We have to support opening a file with zero "Values" defined (as the User could have done something daft in a text editor etc) but we can prevent them from creating such a scenario with our editor.

Fair point. Fair play.

**Comment by tiagobento** [comment]
@manstis AFAIK, some components in PatternFly are "OUIA-Compliant" [1], and some are not. Those which are, have the `ouiaId` property, that not only accepts custom values, but also has a default generated value. That's at least my understanding :)

[1] https://www.patternfly.org/v4/developer-resources/open-ui-automation/

**Review by tiagobento — APPROVED** [review]

**Review by kelvah — APPROVED** [review]

**Pull request merged**